synapse/handlers/message.py

```diff
@@ -969,7 +972,13 @@ async def create_new_client_event(
             event.internal_metadata.outlier = True
             context = EventContext.for_outlier()
         else:
-            context = await self.state.compute_event_context(event)
+            old_state = None
+            # Define the state for historical messages while we know to get all of
+            # state_groups setup properly when we `compute_event_context`.
+            if builder.internal_metadata.is_historical() and full_state_ids_at_event:
+                old_state = await self.store.get_events_as_list(full_state_ids_at_event)
+
+            context = await self.state.compute_event_context(event, old_state=old_state)
```

Comment:
When you pull state from a `state_group`, it grabs the full resolved state from the all of the `state_group` deltas.

So we want to define the state for the historical event as all of the state at that point in time, including member events not matching the `sender` of the event. This way when `/messages` fetches the state for the first event in the batch, it most likely grabs from some historical event with all of the member events necessary for that batch.

This does have the flaw where if you just insert a single event somewhere, it probably won't resolve the state correctly from `/messages` or `/context` since it will grab a non historical event above or below.

Might be good to look at `/mesages` and `/context` to additional add the `state` for any historical messages in that batch.

Reply:
Sharing `state_groups` and resolving state in v2, https://github.com/matrix-org/synapse/pull/10975